receiver/prometheusexec/receiver.go

```diff
@@ -0,0 +1,306 @@
+func (wrapper *prometheusReceiverWrapper) manageProcess() {
+	var (
+		elapsed       time.Duration
+		newPort       int
+		crashCount    int
+		err           error
+		subprocessErr error
+	)
+
+	for {
+
+		// Generate a port if none was specified and if process is unhealthy (Receiver has been stopped)
+		if wrapper.subprocessConfig.Port == 0 && elapsed <= subprocessmanager.HealthyProcessTime {
+			newPort, err = wrapper.handleNewPort(newPort)
+			if err != nil {
+				wrapper.logger.Info("handleNewPort() error - killing this single process/receiver", zap.String("process custom name", wrapper.subprocessConfig.CustomName), zap.String("error", err.Error()))
+				return
+			}
+		}
+
+		// Replace the templating in the strings of the process data
+		wrapper.subprocessConfig = wrapper.fillPortPlaceholders(newPort)
+
+		// Start the receiver if it's the first pass, or if the process is unhealthy and Receiver was stopped last pass
+		if elapsed == 0 || elapsed <= subprocessmanager.HealthyProcessTime && crashCount > subprocessmanager.HealthyCrashCount || elapsed <= subprocessmanager.HealthyProcessTime && wrapper.subprocessConfig.Port == 0 {
+			err = wrapper.prometheusReceiver.Start(wrapper.context, wrapper.host)
+			if err != nil {
+				wrapper.logger.Info("Start() error, could not start receiver - killing this single process/receiver", zap.String("process custom name", wrapper.subprocessConfig.CustomName), zap.String("error", err.Error()))
+				return
+			}
+		}
+
+		// Start the process, error is handled a little later
+		elapsed, subprocessErr = wrapper.subprocessConfig.Run(wrapper.logger)
+
+		// Compute the crashCount depending on subprocess health
+		crashCount, err = wrapper.computeHealthAndCrashCount(elapsed, crashCount)
+		if err != nil {
+			wrapper.logger.Info("computeHealthAndCrashCount() error - killing this single process/receiver", zap.String("process custom name", wrapper.subprocessConfig.CustomName), zap.String("error", err.Error()))
+			return
+		}
+
+		// Compute how long this process will wait before restarting
+		sleepTime := subprocessmanager.GetDelay(elapsed, crashCount)
+
+		// Now we can log the process error, no need to escalate the error, simply log and restart the subprocess
+		if subprocessErr != nil {
+			wrapper.logger.Info("Subprocess error", zap.String("process custom name", wrapper.subprocessConfig.CustomName), zap.String("time until process restarts", sleepTime.String()), zap.String("error", subprocessErr.Error()))
+		}
+
+		// Sleep this goroutine for a certain amount of time, computed by exponential backoff
+		time.Sleep(sleepTime)
+	}
+}
+
+// computeHealthAndCrashCount will decide whether the process is healthy, set the crashCount accordingly and shutdown the receiver if unhealthy
+func (wrapper *prometheusReceiverWrapper) computeHealthAndCrashCount(elapsed time.Duration, crashCount int) (int, error) {
```

Comment:
The return error value is in case the `Shutdown()` method fails for some reason, then I want to propagate that error up the chain to stop that current subprocess/receiver pair!

The *health* part of the name is because if the process is indeed unhealthy, I need to Shutdown its receiver. I'm still open to changing the name of the method if you'd like!